src/compiler/checker.ts

```diff
@@ -15666,11 +15674,22 @@ namespace ts {
             return !!(getGenericObjectFlags(type) & ObjectFlags.IsGenericIndexType);
         }
 
+        function isUnionWithGenericDiscriminantProperty(type: UnionType) {
+            getPropertiesOfType(type); // initialize property cache (includes partial props)
+            const props = arrayFrom((type.propertyCache || emptySymbols).values());
+            for (const prop of props) {
+                if (isDiscriminantProperty(type, prop.escapedName) === DiscriminantKind.Potential) {
+                    return true;
+                }
+            }
+        }
+
         function getGenericObjectFlags(type: Type): ObjectFlags {
             if (type.flags & TypeFlags.UnionOrIntersection) {
                 if (!((type as UnionOrIntersectionType).objectFlags & ObjectFlags.IsGenericTypeComputed)) {
                     (type as UnionOrIntersectionType).objectFlags |= ObjectFlags.IsGenericTypeComputed |
-                        reduceLeft((type as UnionOrIntersectionType).types, (flags, t) => flags | getGenericObjectFlags(t), 0);
+                        reduceLeft((type as UnionOrIntersectionType).types, (flags, t) => flags | getGenericObjectFlags(t), 0) |
+                        (type.flags & TypeFlags.Union && isUnionWithGenericDiscriminantProperty(type as UnionType) ? ObjectFlags.IsGenericObjectType : 0);

```

Comment:
I think it would be easier to read if you flipped the line order. I got confused about whether the new line was inside the `reduceLeft` call or not.
```suggestion
                        (type.flags & TypeFlags.Union && isUnionWithGenericDiscriminantProperty(type as UnionType) ? ObjectFlags.IsGenericObjectType : 0) |
                        reduceLeft((type as UnionOrIntersectionType).types, (flags, t) => flags | getGenericObjectFlags(t), 0);
```